### DISCUSSION TITLE
How to read `AuthenticationProviderKey`, `DownstreamPathTemplate`, `DownstreamHostAndPorts` based on the `UpstreamPath` as input?

### DISCUSSION THREAD

**Discussion by pradeep1403** [ThreeMammals/Ocelot#1987] [Q&A] [answered]
My route information in the **ocelot.json** as 
```json
{
  "DownstreamPathTemplate": "/api/inbound/{RestRoute}",
  "DownstreamScheme": "http"
  "DownstreamHostAndPorts": [
	{ "Host": "Localhost", "Port": 51712 }
  ],
  "AuthenticationOptions":{
	"AuthenticationProviderKey": "ApiKey",
	"AlLowedScopes": []
  },
  "UpstreamPathTemplate": "/api/dev/inbound/{RestRoute}",
  "UpstreamHttpMethod": [ "post","get", "Options"]
}
```
I want to get the `AuthenticationProviderKey`, `DownstreamPathTemplate` and `DownstreamHostAndPorts` based on the `UpstreamPath` (`HttpContext` as in the request). I am using the middleware to capture these information.

### Could you please suggest the way that how can I get the above information when we call the upstream URL?

**Comment by raman-m** [comment, marked as answer]
Hi @pradeep1403 !
Welcome to Ocelot world! 🐯 

In what place/class/middleware do you need properties of the route?

Anyway, route information can be found like this:
- Read `HttpContext.Items` helpers (see [HttpItemsExtensions](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Middleware/HttpItemsExtensions.cs) class). So, you need to use the couple of these methods: `UpsertDownstreamRoute` and `DownstreamRoute`. Look into the Ocelot code and you will see [tons of examples](https://github.com/search?q=repo%3AThreeMammals%2FOcelot+httpContext.Items.DownstreamRoute%28%29&type=code) how to use these extensions
- Use `IOptions<T>` pattern to inject the config model into a constructor. The 1st example of `IOptions<FileConfiguration>` usage is here:
 
 https://github.com/ThreeMammals/Ocelot/blob/5a3c55e358afb7853819e3370daad9bb15452dd9/src/Ocelot.Provider.Consul/ConsulFileConfigurationRepository.cs#L21

So, the 2nd dev approach gives you what you want, but you have to find required route by a predicate. It is a bit overhead coding! Just use 1st approach with concrete `DownstreamRoute` obj detection and propagate it to your class/context.

Hope it helps!
